### ISSUE TITLE
BUG: save a float32 array and then load it gives different arrays!

### ISSUE THREAD

**Issue by ForeverZyh** [numpy/numpy#25175]
### Describe the issue:

```
>>> import numpy as np
>>> import torch
>>> load_x = torch.load("data_torch").numpy()
>>> np.sum(load_x)
1452856.1
>>> load_x.dtype
dtype('float32')
>>> np.save("tmp", load_x, allow_pickle=True)
>>> load_x = np.load("tmp.npy", allow_pickle=True)
>>> np.sum(load_x)
1452855.4
>>> load_x.dtype
dtype('float32')
```

I saved a file (around 12MB) onto the disk and then immediately loaded it. They gave me different answers. I tried different disks so that is not related to my disks. 
Actually, if I save a 800MB file and load it, the error will become even larger.

### Reproduce the code example:

```python
import numpy as np
import torch
load_x = torch.load("data_torch").numpy()
np.sum(load_x)
load_x.dtype
np.save("tmp", load_x, allow_pickle=True)
load_x = np.load("tmp.npy", allow_pickle=True)
np.sum(load_x)
load_x.dtype
```

### Runtime information:

1.24.3
3.8.18 (default, Sep 11 2023, 13:20:55) 
[GCC 11.2.0]

In fact, I tried 1.22.0, 1.22.4, 1.23.0 etc. None of them work. And I can reproduce it on MacOS and Linux.

### Context for the issue:

I use torch to just load the file because I cannot save it using numpy...
If you want to have download the torch file, please see here https://drive.google.com/file/d/1f1MbFKktjSTthjhwhCKcUX1XbbJn19h_/view?usp=sharing

**Comment by rkern** [comment]
The values are precisely the same. The only difference is that the first `load_x` from `torch` is not C-contiguous. However, it gets saved in C-contiguous order when written out to NPY and then loaded again. This is generally considered fair-game.

`np.sum()` does an optimization in the order in which it does the summations when it is given an array that is not C-contiguous. It will try to find the axis that is contiguous and do the summations there first. Reordering computations to happen along the "fast" axis is also generally considered fair-game. Floating point arithmetic _is_ sensitive to order, and the differences that you see are within the expected numerical error. Both answers are plausible `float32` answers to what is the sum of the `float32` numbers in your array. Other reorderings will give you similarly-variable answers, all valid.

There's an argument to be made that `np.sum()` should operate in the same order of elements as exposed to Python indexing, regardless of the internal memory ordering as a way to avoid surprises like this. But it's a design choice, not a bug. I am going to close this issue, as I don't think we are going to make changes to numpy in response to this. Please feel free to continue to discuss here on the closed issue if you have more questions. numpy devs, if there _are_ any changes you think you want to take up in response to this, feel free to reopen.

**Comment by ForeverZyh** [comment]
Thanks for your reply. Yes, they are the same as using `np.max(np.abs(a-b))` gives me zero. 

[This doc gives really good explanation](https://numpy.org/doc/stable/reference/generated/numpy.sum.html)
```
For floating point numbers the numerical precision of sum (and np.add.reduce) is in general limited by directly adding each number individually to the result causing rounding errors in every step. However, often numpy will use a numerically better approach (partial pairwise summation) leading to improved precision in many use-cases. This improved precision is always provided when no axis is given. When axis is given, it will depend on which axis is summed. Technically, to provide the best speed possible, the improved precision is only used when the summation is along the fast axis in memory. 
```

So I think a quick workaround for me is just using `np.sum(dtype='float64')`.
